### PR TITLE
[SNOW-1541092] Add RepeatedSubqueryElimination for the new compilation stage

### PR DESCRIPTION
**Pull request by sfc-gh-yzou**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   Fixes SNOW-1541092

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.

3. Please describe how your code solves the related issue.

 Move the repeated subquery elimination to the query compilation stage with plan transformation:
1) Introduced new logical plan node WithQueryBlock, and propagated the referred cte through plan building.
2) Added plan transformation once duplicated node is detected.
3) Handle query generation for WithQueryBlock

### CODE REVIEW

File: src/snowflake/snowpark/_internal/analyzer/select_statement.py [hunk 1]
```diff
@@ -373,6 +375,11 @@ def column_states(self, value: ColumnStateDict):
         """
         self._column_states = deepcopy(value)
 
+    @abstractmethod
+    def referred_cte_tables(self) -> Set[str]:
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/_internal/analyzer/select_statement.py:379]:
```suggestion
    def reference_ctes(self) -> Set[str]:
```
I guess this name is more accurate?

File: src/snowflake/snowpark/_internal/analyzer/select_statement.py [hunk 2]
```diff
@@ -1261,6 +1292,14 @@ def individual_node_complexity(self) -> Dict[PlanNodeCategory, int]:
         # we add #set_operands - 1 additional operators in sql query
         return {PlanNodeCategory.SET_OPERATION: len(self.set_operands) - 1}
 
+    def referred_cte_tables(self) -> Set[str]:
+        # get a union of referred cte tables from all child nodes
+        referred_cte_tables: Set[str] = set()
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/_internal/analyzer/select_statement.py:1297]: maybe
```
return set().union(*[node.referred_cte_tables() for node in self._nodes])
```

File: src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py [hunk 3]
```diff
@@ -222,6 +224,10 @@ def __init__(
         # TODO (SNOW-1541096): Remove placeholder_query once CTE is supported with the
         #               new compilation step.
         placeholder_query: Optional[str] = None,
+        # This field records all the CTE tables that are referred by the
+        # current SnowflakePlan tree. This is needed for the final query
+        # generation to generate the correct sql query with CTE definition.
+        referred_cte_tables: Optional[Set[str]] = None,
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py:230]: does order matter here? (I think should?) Set is not ordered in python

File: src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py [hunk 4]
```diff
@@ -511,6 +526,7 @@ def build(
         source_plan: Optional[LogicalPlan],
         schema_query: Optional[str] = None,
         is_ddl_on_temp_object: bool = False,
+        propagate_referred_cte_tables: bool = True,
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py:529]: can you add a comment about when it should be true or false?

File: src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py [hunk 5]
```diff
@@ -601,15 +606,66 @@ def build_binary(
         }
         api_calls = [*select_left.api_calls, *select_right.api_calls]
 
+        # This is a temporary workaround for query comparison. The previous CTE optimization
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py:609]: You mean this? https://github.com/snowflakedb/snowpark-python/blob/89913c27ed588e0f49ba729cb9f85b835a17bf2d/src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py#L1456

this is not introduced by CTE, and it's used for retrieving result from previous query id https://github.com/snowflakedb/snowpark-python/blob/89913c27ed588e0f49ba729cb9f85b835a17bf2d/src/snowflake/snowpark/_internal/server_connection.py#L644-L646

you can follow this https://github.com/snowflakedb/snowpark-python/blob/89913c27ed588e0f49ba729cb9f85b835a17bf2d/src/snowflake/snowpark/_internal/analyzer/cte_utils.py#L148 to compare sql query text and query_params. Maybe you can implement __eq__ in Query so you directly use `if current_query in existing_queries`

File: src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py [hunk 6]
```diff
@@ -601,15 +606,66 @@ def build_binary(
...
+        referred_cte_tables: Set[str] = set()
+        if self.session.cte_optimization_enabled and is_new_compilation_stage_enabled():
+            # When the cte optimization and the new compilation stage is enabled, the
+            # queries, referred cte tables, and post actions propogated from
+            # left and right can have duplicated queries if there is a common CTE block referred.
+            # Need to do a deduplication to avoid repeated query.
+            merged_queries = select_left.queries[:-1].copy()
+            for query in select_right.queries[:-1]:
+                if not _query_exists(query, merged_queries):
+                    merged_queries.append(query)
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py:634]: also copy here?

File: src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py [hunk 7]
```diff
@@ -601,15 +606,66 @@ def build_binary(
...
+
+            referred_cte_tables.update(select_left.referred_cte_tables)
+            referred_cte_tables.update(select_right.referred_cte_tables)
+
+            post_actions = select_left.post_actions.copy()
+            for post_action in select_right.post_actions:
+                if post_action not in post_actions:
+                    post_actions.append(post_action)
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py:642]: also copy here?

File: src/snowflake/snowpark/_internal/compiler/common_subdataframe_elimination.py [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+        self._query_generator = query_generator
+        self._node_count_map = defaultdict(int)
+        self._node_parents_map = defaultdict(set)
+        self._duplicated_nodes = set()
+
+    def apply(self) -> List[LogicalPlan]:
+        """
+        Applies Common SubDataframe elimination on the set of logical plans one after another.
+
+        Returns
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/_internal/compiler/common_subdataframe_elimination.py:76]:
```suggestion
        Returns:
```
and remove dash lines, and adding indent to the following line, which is consistent inside snowpark python code

File: src/snowflake/snowpark/_internal/compiler/common_subdataframe_elimination.py [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+            if len(self._duplicated_nodes) > 0:
+                deduplicated_plan = self._deduplicate_with_cte(logical_plan)
+                final_logical_plans.append(deduplicated_plan)
+            else:
+                final_logical_plans.append(logical_plan)
+
+        # TODO (SNOW-1566363): Add telemetry for CTE
+        return final_logical_plans
+
+    def _find_duplicate_subtrees(self, root: TreeNode) -> None:
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/_internal/compiler/common_subdataframe_elimination.py:105]: Is it same with the implementaion in cte_utils.py? If yes, maybe import from it first?

File: src/snowflake/snowpark/_internal/compiler/utils.py [hunk 10]
```diff
@@ -103,49 +107,113 @@ def to_selectable(plan: LogicalPlan, analyzer: Analyzer) -> Selectable:
...
+    #                  |
+    #                JOIN
+    # when reset the child of join, and parent is SelectSnowflakePlan, we make sure
+    # all nodes along the way up to SelectSnowflakePlan are reset correctly.
+    reset_node(parent)
+
+
+def reset_node(node: LogicalPlan) -> None:
+    """
+    Helper methods that is used to clean up cached query filed of the given node
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/_internal/compiler/utils.py:167]:
```suggestion
    Helper method that is used to clean up cached query filed of the given node
```
what's cached_query?

File: src/snowflake/snowpark/_internal/compiler/common_subdataframe_elimination.py [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+        for parent in parents:
+            replace_child_and_reset_node(parent, node, new_node, self._query_generator)
+
+    def _resolve_node(self, node: LogicalPlan) -> SnowflakePlan:
+        resolved_node = self._query_generator.resolve(node)
+        resolved_node._is_valid_for_replacement = True
+
+        return resolved_node
+
+    def _deduplicate_with_cte(self, root: "TreeNode") -> LogicalPlan:
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/_internal/compiler/common_subdataframe_elimination.py:176]: I'd call it `_replace_duplicate_node_with_cte`, which is clearer

File: src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py [hunk 12]
```diff
@@ -1432,8 +1499,35 @@ def add_result_scan_if_not_select(self, plan: SnowflakePlan) -> SnowflakePlan:
...
+
+    def with_query_block(
+        self, name: str, child: SnowflakePlan, source_plan: LogicalPlan
+    ) -> SnowflakePlan:
+        if self._skip_schema_query is not True:
+            raise ValueError(
+                "schema query for WithQueryBlock is currently not supported"
             )
 
+        new_query = f"SELECT * FROM {name}"
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py:1513]: can you use `project_statement` in analyzer_utils.py

File: src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py [hunk 13]
```diff
@@ -1432,8 +1499,35 @@ def add_result_scan_if_not_select(self, plan: SnowflakePlan) -> SnowflakePlan:
                 plan.source_plan,
                 api_calls=plan.api_calls,
                 session=self.session,
+                referred_cte_tables=plan.referred_cte_tables,
+            )
+
+    def with_query_block(
+        self, name: str, child: SnowflakePlan, source_plan: LogicalPlan
+    ) -> SnowflakePlan:
+        if self._skip_schema_query is not True:
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py:1508]:
```suggestion
        if not self._skip_schema_query:
```

File: src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py [hunk 14]
```diff
@@ -1432,8 +1499,35 @@ def add_result_scan_if_not_select(self, plan: SnowflakePlan) -> SnowflakePlan:
...
+            raise ValueError(
+                "schema query for WithQueryBlock is currently not supported"
             )
 
+        new_query = f"SELECT * FROM {name}"
+
+        queries = child.queries[:-1] + [Query(sql=new_query)]
+        # propogate the cte table
+        referred_cte_tables = {name}
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py:1517]:
```suggestion
        referred_cte_tables = {name}.union(child.referred_cte_tables)
```

File: src/snowflake/snowpark/_internal/compiler/plan_compiler.py [hunk 15]
```diff
@@ -48,13 +57,24 @@ def should_apply_optimizations(self) -> bool:
         ) and current_session.cte_optimization_enabled
 
     def compile(self) -> Dict[PlanQueryType, List[Query]]:
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/_internal/compiler/plan_compiler.py:59]: this method will not be triggered by any existing code path, right? (outside of query optimization)

File: src/snowflake/snowpark/_internal/compiler/utils.py [hunk 16]
```diff
@@ -103,49 +107,113 @@ def to_selectable(plan: LogicalPlan, analyzer: Analyzer) -> Selectable:
...
+    The reset operation is only needed for resolved node like Selectable or SnowflakePlan,
+    since only those two types of nodes have cached queries.
+    """
+
+    def reset_selectable(selectable_node: Selectable) -> None:
+        # clear the cache to allow it re-calculate the _snowflake_plan and _sql_query
+        if not isinstance(node, (SelectSnowflakePlan, SelectSQL, SelectTableFunction)):
+            selectable_node._snowflake_plan = None
+        if isinstance(node, (SelectStatement, SetStatement)):
+            selectable_node._sql_query = None
```
Comment (sfc-gh-aalam) [inline src/snowflake/snowpark/_internal/compiler/utils.py:179]: for `SetStatement` you should also reset plan for the `set_operand.selectables`

File: src/snowflake/snowpark/_internal/compiler/plan_compiler.py (hunk 15 above)
Comment (sfc-gh-yzou) [reply src/snowflake/snowpark/_internal/compiler/plan_compiler.py:59]: No, this method is already triggered today for execution_queries in general, however, under the case when no optimization is enabled, it is doing the same as today.

File: src/snowflake/snowpark/context.py [hunk 17]
```diff
@@ -20,6 +20,18 @@
 # If _should_continue_registration is not None, i.e. a caller environment has assigned it an alternate callable, then the callback is responsible for determining the rest of the Snowpark workflow.
 _should_continue_registration: Optional[Callable[..., bool]] = None
 
+# If true, the final new
+_enable_new_compilation_stage = False
```
[the remark this answers is not in the document]
Comment (sfc-gh-yzou) [reply src/snowflake/snowpark/context.py]: i actually decided to move it to session parameter, i think it would be good to have a global parameter to turn everything off.

File: tests/integ/test_cte.py [hunk 18]
```diff
@@ -64,7 +76,7 @@ def check_result(session, df, expect_cte_optimized):
     if installed_pandas:
         from pandas.testing import assert_frame_equal
 
-        assert_frame_equal(result_pandas, cte_result_pandas)
+        assert_frame_equal(result_pandas, cte_result_pandas, check_dtype=False)
```
[the remark this answers is not in the document]
Comment (sfc-gh-yzou) [reply tests/integ/test_cte.py]: not needed, removed

File: src/snowflake/snowpark/_internal/compiler/common_subdataframe_elimination.py (hunk 8 above)
Comment (sfc-gh-yzou) [reply src/snowflake/snowpark/_internal/compiler/common_subdataframe_elimination.py]: updated

File: src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py (hunk 14 above)
Comment (sfc-gh-yzou) [reply src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py]: updated

File: src/snowflake/snowpark/_internal/compiler/utils.py (hunk 16 above)
Comment (sfc-gh-yzou) [reply src/snowflake/snowpark/_internal/compiler/utils.py]: There is no reset anymore

File: src/snowflake/snowpark/_internal/analyzer/select_statement.py [hunk 19]
```diff
@@ -373,6 +374,11 @@ def column_states(self, value: ColumnStateDict):
         """
         self._column_states = deepcopy(value)
 
+    @abstractmethod
+    def referred_ctes(self) -> Set[str]:
+        """Return the set of ctes referred by the current selectable node"""
```
Comment (sfc-gh-aalam) [inline src/snowflake/snowpark/_internal/analyzer/select_statement.py:379]: is this the set of CTE's that are referenced inside the subtree rooted at this node? can we update the description to include we mean subtree here.

File: src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py [hunk 20]
```diff
@@ -601,15 +608,72 @@ def build_binary(
...
+            referred_ctes.update(select_left.referred_ctes)
+            referred_ctes.update(select_right.referred_ctes)
+
+            post_actions = select_left.post_actions.copy()
+            for post_action in select_right.post_actions:
+                if post_action not in post_actions:
+                    post_actions.append(copy.copy(post_action))
+        else:
+            merged_queries = select_left.queries[:-1] + select_left.queries[:-1]
+            post_actions = select_left.post_actions + select_right.post_actions
```
Comment (sfc-gh-aalam) [inline src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py:653]:
```suggestion
            merged_queries = select_left.queries[:-1] + select_right.queries[:-1]
            post_actions = select_left.post_actions + select_right.post_actions
```

File: src/snowflake/snowpark/_internal/compiler/plan_compiler.py [hunk 21]
```diff
@@ -29,11 +34,12 @@ class PlanCompiler:
...
         self._plan = plan
 
-    def should_apply_optimizations(self) -> bool:
+    def should_start_query_compilation(self) -> bool:
         """
         Whether optimization should be applied to the plan or not.
         Optimization can be applied if
         1) there is source logical plan attached to the current snowflake plan
+        2) the query compilation stage is enabled
         2) optimizations are enabled in the current session, such as cte_optimization_enabled
```
Comment (sfc-gh-aalam) [inline src/snowflake/snowpark/_internal/compiler/plan_compiler.py:43]:
```suggestion
        2) the query compilation stage is enabled
        3) optimizations are enabled in the current session, such as cte_optimization_enabled
```

File: src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py [hunk 22]
```diff
@@ -1432,8 +1507,34 @@ def add_result_scan_if_not_select(self, plan: SnowflakePlan) -> SnowflakePlan:
...
+        if not self._skip_schema_query:
+            raise ValueError(
+                "schema query for WithQueryBlock is currently not supported"
             )
 
+        new_query = project_statement([], name)
+
+        queries = child.queries[:-1] + [Query(sql=new_query)]
+        # propogate the cte table
+        referred_ctes = {name}.union(child.referred_ctes)
```
Comment (sfc-gh-aalam) [inline src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py:1525]:
```suggestion
        # propagate the cte table
        referred_ctes = {name}.union(child.referred_ctes)
```

File: src/snowflake/snowpark/_internal/compiler/plan_compiler.py [hunk 23]
```diff
@@ -44,17 +50,33 @@ def should_apply_optimizations(self) -> bool:
...
+            logical_plans = [copy.deepcopy(self._plan)]
+            # 2. create a code generator with the original plan
+            query_generator = create_query_generator(self._plan)
+
+            # apply each optimizations if needed
+            if self._plan.session.cte_optimization_enabled:
+                common_sub_dataframe_eliminator = RepeatedSubqueryElimination(
+                    logical_plans, query_generator
+                )
+                logical_plans = common_sub_dataframe_eliminator.apply()
```
Comment (sfc-gh-aalam) [inline src/snowflake/snowpark/_internal/compiler/plan_compiler.py:71]: nit: rename the object?
```suggestion
                repeated_subquery_eliminator = RepeatedSubqueryElimination(
                    logical_plans, query_generator
                )
                logical_plans = repeated_subquery_eliminator.apply()
```

File: src/snowflake/snowpark/_internal/analyzer/select_statement.py (hunk 19 above)
Comment (sfc-gh-yzou) [reply src/snowflake/snowpark/_internal/analyzer/select_statement.py:379]: yes, added

File: src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py (hunk 20 above)
Comment (sfc-gh-yzou) [reply src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py:654]: Oh, good catch

File: src/snowflake/snowpark/_internal/compiler/plan_compiler.py (hunk 21 above)
Comment (sfc-gh-yzou) [reply src/snowflake/snowpark/_internal/compiler/plan_compiler.py:43]: updated

File: src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py (hunk 22 above)
Comment (sfc-gh-yzou) [reply src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py]: updated

File: src/snowflake/snowpark/_internal/compiler/plan_compiler.py (hunk 23 above)
Comment (sfc-gh-yzou) [reply src/snowflake/snowpark/_internal/compiler/plan_compiler.py]: updated

File: src/snowflake/snowpark/_internal/compiler/utils.py [hunk 24]
```diff
@@ -58,26 +62,71 @@ def create_query_generator(plan: SnowflakePlan) -> QueryGenerator:
...
+
+def resolve_node(node: LogicalPlan, query_generator: QueryGenerator) -> SnowflakePlan:
+    resolved_node = query_generator.resolve(node)
+    resolved_node._is_valid_for_replacement = True
+
+    return resolved_node
+
+
+def replace_child_and_update_parent(
     parent: LogicalPlan,
```
Comment (sfc-gh-aalam) [inline src/snowflake/snowpark/_internal/compiler/utils.py:100]: I think we should keep replace_child and update parent snowflake plan separate. The reason that they are together now is because we don't want to keep the snowflake plan to have inconsistent queries. But in this function we are not completely addressing the inconsistency problem. Once parent is updated with correct queries, we need to also update the parent's parent.

Since that will be done separately anyways, what's the advantage of combining the two functions here?

File: tests/conftest.py [hunk 25]
```diff
@@ -27,7 +27,7 @@ def is_excluded_frontend_file(path):
 
 def pytest_addoption(parser):
     parser.addoption("--disable_sql_simplifier", action="store_true", default=False)
-    parser.addoption("--local_testing_mode", action="store_true", default=False)
+    parser.addoption("--local_testing_mode", action="store_true", default=True)
```
Comment (sfc-gh-aalam) [inline tests/conftest.py:30]: why is this change here?

File: src/snowflake/snowpark/session.py [hunk 26]
```diff
@@ -194,9 +194,12 @@
...
     "PYTHON_SNOWPARK_USE_LOGICAL_TYPE_FOR_CREATE_DATAFRAME"
 )
+# parameter used to turn off the whole new query compilation stage in one shot. If turned
+# off the plan won't go through the extra optimization and query generation steps.
+_PYTHON_SNOWPARK_ENABLE_QUERY_COMPILATION_STAGE = (
+    "PYTHON_SNOWPARK_ENABLE_COMPILATION_STAGE"
+)
 _PYTHON_SNOWPARK_USE_CTE_OPTIMIZATION_STRING = "PYTHON_SNOWPARK_USE_CTE_OPTIMIZATION"
-# TODO (SNOW-1482588): Add parameter for PYTHON_SNOWPARK_ELIMINATE_NUMERIC_SQL_VALUE_CAST_ENABLED
-#               at server side
```
Comment (sfc-gh-aalam) [inline src/snowflake/snowpark/session.py:199]: is this done?

Comment (sfc-gh-yzou) [reply src/snowflake/snowpark/session.py:199]: yes

File: tests/conftest.py (hunk 25 above)
Comment (sfc-gh-yzou) [reply tests/conftest.py:30]: reverted, i was runing test locally with local testing mode

File: src/snowflake/snowpark/_internal/compiler/utils.py (hunk 24 above)
Comment (sfc-gh-yzou) [reply src/snowflake/snowpark/_internal/compiler/utils.py:100]: this is trying to make the node consistent from parent to child, the benefit of making them together is that whenever child is updated, the corresponding parent node needs to be updated, this is to avoid we forgot to update anything. 
Do you have a use case that you want to handle them separately? maybe we can refactor when you need this, which can make things little bit cleaner about why you need this?

File: src/snowflake/snowpark/_internal/compiler/plan_compiler.py [hunk 27]
```diff
@@ -44,17 +50,33 @@ def should_apply_optimizations(self) -> bool:
 
         current_session = self._plan.session
         return (
-            self._plan.source_plan is not None
-        ) and current_session.cte_optimization_enabled
+            (self._plan.source_plan is not None)
+            and current_session.query_compilation_stage_enabled
+            and current_session.cte_optimization_enabled
+        )
```
[the remark this answers is not in the document]
Comment (sfc-gh-yzou) [reply src/snowflake/snowpark/_internal/compiler/plan_compiler.py:57]: why is that? we do need to go through the pre-process steps like copy if no optimization needs to be applied

[the remark this answers is not in the document]
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/compiler/plan_compiler.py:57]: I see. Then I guess I update this to say `current_session.cte_optimization_enabled or current_session.large_query_breakdown_enabled`

File: src/snowflake/snowpark/_internal/compiler/utils.py (hunk 24 above)
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/compiler/utils.py:100]: in repeated subquery elimination you update ancestors using `_update_parents` in a bottom-up traversal. When a with_query block is added we call `self._update_parents(node, resolved_with_block)` to make the child replacement and updating the queries and mark the parents to be updated later. Then for the updated parents, we call `self._update_parents(node, node)`. 

I think it would be cleaner if we do something like
```
            if node in self._duplicated_nodes:
                # create a WithQueryBlock node
                with_block = WithQueryBlock(
                    name=random_name_for_temp_object(TempObjectType.CTE), child=node
                )
                with_block._is_valid_for_replacement = True

                resolved_with_block = resolve_node(with_block, self._query_generator)
                replace_child(node, resolved_with_block)
                self._update_parents(node)
            elif node in self._updated_nodes:
                # if the node is updated, make sure all nodes up to parent is updated
                self._update_parents(node)
```

in large query breakdown, I am following a similar pattern. We can still use `replace_child_and_update_parent(node, node)` we aren't really replacing a child. Only using it to re-resolve parent queries.

Comment (sfc-gh-jdu) [reply src/snowflake/snowpark/_internal/compiler/utils.py:100]: agree

File: src/snowflake/snowpark/_internal/compiler/utils.py [hunk 28]
```diff
@@ -58,26 +62,71 @@ def create_query_generator(plan: SnowflakePlan) -> QueryGenerator:
...
+        new_snowflake_plan.df_aliased_col_name_to_real_col_name
+    )
+    node.placeholder_query = new_snowflake_plan.placeholder_query
+    node.referred_ctes = new_snowflake_plan.referred_ctes
+    node._cumulative_node_complexity = new_snowflake_plan._cumulative_node_complexity
+
+
+def resolve_node(node: LogicalPlan, query_generator: QueryGenerator) -> SnowflakePlan:
+    resolved_node = query_generator.resolve(node)
+    resolved_node._is_valid_for_replacement = True
```
Comment (sfc-gh-aalam) [inline src/snowflake/snowpark/_internal/compiler/utils.py:94]: should we add this inside query_generator since we know this snowflakeplan will be generated during optimization stage?

File: CHANGELOG.md [hunk 29]
```diff
@@ -6,6 +6,7 @@
 
 #### New Features
 - Added support for `snowflake.snowpark.testing.assert_dataframe_equal` that is a util function to check the equality of two Snowpark DataFrames.
+- Added query compilation stage with extra optimization and query generation step. The stage can be enabled by setting `session.query_compilation_stage_enabled = True`.
```
[the remark this answers is not in the document]
Comment (sfc-gh-yzou) [reply CHANGELOG.md]: customer theoretically don't have to know, let me revert that for now

File: src/snowflake/snowpark/_internal/analyzer/select_statement.py [hunk 30]
```diff
@@ -373,6 +374,11 @@ def column_states(self, value: ColumnStateDict):
         """
         self._column_states = deepcopy(value)
 
+    @abstractmethod
+    def referred_ctes(self) -> Set[str]:
```
[the remark this answers is not in the document]
Comment (sfc-gh-yzou) [reply src/snowflake/snowpark/_internal/analyzer/select_statement.py]: yeah, I meant *referenced*

File: src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py (hunk 5 above)
Comment (sfc-gh-yzou) [reply src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py]: sure, updated

File: src/snowflake/snowpark/_internal/compiler/repeated_subquery_elimination.py [hunk 31]
```diff
@@ -0,0 +1,153 @@
...
+
+        Returns:
+            A set of the new LogicalPlans with common sub dataframe deduplicated with CTE node.
+        """
+        final_logical_plans: List[LogicalPlan] = []
+        for logical_plan in self._logical_plans:
+            # clear the node_count and parents map
+            self._node_count_map = defaultdict(int)
+            self._node_parents_map = defaultdict(set)
+            self._duplicated_nodes.clear()
```
[the remark this answers is not in the document]
Comment (sfc-gh-yzou) [reply src/snowflake/snowpark/_internal/compiler/repeated_subquery_elimination.py]: I removed all those book keeping node. Originally, I was thinking about recording a gloabl state for telemetry usage, but we can do it later if needed

File: src/snowflake/snowpark/_internal/compiler/repeated_subquery_elimination.py [hunk 32]
```diff
@@ -0,0 +1,153 @@
...
+        parents = self._node_parents_map[node]
+        for parent in parents:
+            replace_child_and_update_parent(
+                parent, node, new_node, self._query_generator
+            )
+            self._updated_nodes.add(parent)
+
+    def _replace_duplicate_node_with_cte(self, root: "TreeNode") -> LogicalPlan:
+        """
+        Replace all duplicated nodes with a WithQueryBlock (CTE node), to enable
```
[the remark this answers is not in the document]
Comment (sfc-gh-yzou) [reply src/snowflake/snowpark/_internal/compiler/repeated_subquery_elimination.py:93]: added

File: src/snowflake/snowpark/_internal/compiler/utils.py (hunk 28 above)
Comment (sfc-gh-yzou) [reply src/snowflake/snowpark/_internal/compiler/utils.py]: that is a good idea, i moved this to query generator resolve_with_children

File: src/snowflake/snowpark/_internal/compiler/utils.py (hunk 24 above)
Comment (sfc-gh-yzou) [reply src/snowflake/snowpark/_internal/compiler/utils.py]: i actually updated it to two function

File: src/snowflake/snowpark/_internal/compiler/plan_compiler.py [hunk 33]
```diff
@@ -29,12 +35,13 @@ class PlanCompiler:
     def __init__(self, plan: SnowflakePlan) -> None:
         self._plan = plan
 
-    def should_apply_optimizations(self) -> bool:
+    def should_start_query_compilation(self) -> bool:
```
[the remark this answers is not in the document]
Comment (sfc-gh-yzou) [reply src/snowflake/snowpark/_internal/compiler/plan_compiler.py:38]: the original name is not accurate, if i added check of current_session.query_compilation_stage_enabled. might prefer this name

File: src/snowflake/snowpark/_internal/compiler/repeated_subquery_elimination.py [hunk 34]
```diff
@@ -0,0 +1,153 @@
...
+)
+
+# common sub-dataframe elimination only operates on top
+TreeNode = Union[SnowflakePlan, Selectable]
+
+
+class RepeatedSubqueryElimination:
+    """
+    Optimization that used eliminate duplicated queries that is generated for the
+    common sub-dataframes.
```
[the remark this answers is not in the document]
Comment (sfc-gh-yzou) [reply src/snowflake/snowpark/_internal/compiler/repeated_subquery_elimination.py]: udpated

File: src/snowflake/snowpark/_internal/compiler/query_generator.py [hunk 35]
```diff
@@ -116,18 +148,65 @@ def do_resolve_with_resolved_children(
...
-            logical_plan, resolved_children, df_aliased_col_name_to_real_col_name
-        )
+        elif isinstance(logical_plan, WithQueryBlock):
+            resolved_child = resolved_children[logical_plan.children[0]]
+            # record the CTE definition of the current block
+            if logical_plan.name not in self.resolved_with_query_block:
+                self.resolved_with_query_block[
+                    logical_plan.name
+                ] = resolved_child.queries[-1].sql
+
```
Comment (sfc-gh-aalam) [inline src/snowflake/snowpark/_internal/compiler/query_generator.py:198]: I'm curious to see if this works well for nested CTE - case where one block contains another block. do we have any test for that?

**Review by sfc-gh-aalam — APPROVED** [review]
looks good. Just want to make sure that nested CTEs are tested

File: src/snowflake/snowpark/_internal/compiler/query_generator.py (hunk 35 above)
Comment (sfc-gh-yzou) [reply src/snowflake/snowpark/_internal/compiler/query_generator.py:198]: yes, @sfc-gh-jdu have a test case for one cte depends on the other cte here https://github.com/snowflakedb/snowpark-python/blob/main/tests/integ/test_cte.py#L139. However, we do not generate nested CTE statement, they are all flattened with proper order, i believe that was due to a performance concern.

Comment (sfc-gh-jdu) [reply src/snowflake/snowpark/_internal/compiler/query_generator.py:198]: yes

File: src/snowflake/snowpark/_internal/analyzer/select_statement.py [hunk 36]
```diff
@@ -373,6 +374,11 @@ def column_states(self, value: ColumnStateDict):
         """
         self._column_states = deepcopy(value)
 
+    @abstractmethod
+    def referenced_ctes(self) -> Set[str]:
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/_internal/analyzer/select_statement.py:378]: maybe make it a property?

**Review by sfc-gh-jdu — APPROVED** [review]

File: src/snowflake/snowpark/session.py [hunk 37]
```diff
@@ -635,6 +644,10 @@ def auto_clean_up_temp_table_enabled(self) -> bool:
         """
         return self._auto_clean_up_temp_table_enabled
 
+    @property
+    def query_compilation_stage_enabled(self) -> bool:
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/session.py:648]: just another thought, we might not need to expose this parameter to users? We can only have `self._query_compilation_stage_enabled`

Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/session.py:648]: yeah. good point

File: src/snowflake/snowpark/_internal/analyzer/select_statement.py (hunk 36 above)
Comment (sfc-gh-yzou) [reply src/snowflake/snowpark/_internal/analyzer/select_statement.py:379]: sure, updated

File: src/snowflake/snowpark/session.py (hunk 37 above)
Comment (sfc-gh-yzou) [reply src/snowflake/snowpark/session.py]: as discussed offline, we removed the client side user facing parameter control for now
